### PR TITLE
feat: Add NexusViewer and AssessmentViewer

### PR DESCRIPTION
**Pull request by rackstar**
## Context

Closes https://github.com/NexusMutual/smart-contracts/issues/1165
Closes https://github.com/NexusMutual/smart-contracts/issues/1166

NOTE: this PR is behind https://github.com/NexusMutual/smart-contracts/pull/1170

## Changes proposed in this pull request

* add AssessmentViewer
  * add AssessmentRewards struct
  * add getRewards
* add NexusViewer
  * getClaimableNXM
  * getStakedNXM 
* update StakingViewer
  * add ManagerPoolsAndRewards struct
  * add getManagedStakingPools
  * add getManagerTokenRewards
  * add getManagerPoolsAndRewards
* add IStakingViewer
* add IAssessmentViewer
* deployments
  * drop deprecated contracts
  * add, CoverProducts, StakingViewer, AssessmentViewer and NexusViewer

## Deployment

Adjust baseFee (`-b XX`) according to [current network baseFee](https://etherscan.io/gastracker)

AssessmentViewer - [0xcafeaC01F6BA70567f5EA7dD4e114f8E5FDa5bf6](https://dashboard.tenderly.co/NexusMutual/nexusmutual/testnet/1598735e-f0f8-4151-91fe-f006da62969f/contract/virtual/0xcafeaC01F6BA70567f5EA7dD4e114f8E5FDa5bf6)
Constructor Params:
* NXMaster - [0x01BFd82675DBCc7762C84019cA518e701C0cD07e](https://sdk.nexusmutual.io/)
```bash
node scripts/create2/deploy.js AssessmentViewer -a 0xcafeaC01F6BA70567f5EA7dD4e114f8E5FDa5bf6 \
                                                -f 0xfac7011663910F75CbE1E25539ec2D7529f93C3F \
                                                -c '0x01BFd82675DBCc7762C84019cA518e701C0cD07e' \
                                                -s 36024510 
                                                -b XX \
                                                -p 1
```

StakingViewer - [0xcafeaaf26278671fc393c60c09d936de2E3F6a1a](https://dashboard.tenderly.co/NexusMutual/nexusmutual/testnet/1598735e-f0f8-4151-91fe-f006da62969f/contract/virtual/0xcafeaaf26278671fc393c60c09d936de2E3F6a1a)
Constructor Params:
* NXMaster - [0x01BFd82675DBCc7762C84019cA518e701C0cD07e](https://sdk.nexusmutual.io/)
* StakingNFT - [0xcafea508a477D94c502c253A58239fb8F948e97f](https://sdk.nexusmutual.io/)
* StakingPoolFactory - [0xcafeafb97BF8831D95C0FC659b8eB3946B101CB3](https://sdk.nexusmutual.io/)
```bash
node scripts/create2/deploy.js StakingViewer -a 0xcafeaaf26278671fc393c60c09d936de2E3F6a1a \
                                             -f 0xfac7011663910F75CbE1E25539ec2D7529f93C3F \
                                             -c '["0x01BFd82675DBCc7762C84019cA518e701C0cD07e","0xcafea508a477D94c502c253A58239fb8F948e97f","0xcafeafb97BF8831D95C0FC659b8eB3946B101CB3"]' \
                                             -s 116468797 
                                             -b XX \
                                             -p 1
```

NexusViewer - [0xcafeaaaA618809F4414fa9e2eB7Db9975419fa3F](https://dashboard.tenderly.co/NexusMutual/nexusmutual/testnet/1598735e-f0f8-4151-91fe-f006da62969f/contract/virtual/0xcafeaaaa618809f4414fa9e2eb7db9975419fa3f)
Constructor Params:
* NXMaster - [0x01BFd82675DBCc7762C84019cA518e701C0cD07e](https://sdk.nexusmutual.io/)
* StakingViewer - 0xcafeaaf26278671fc393c60c09d936de2E3F6a1a (see above)
* AssessmentViewer - 0xcafeaC01F6BA70567f5EA7dD4e114f8E5FDa5bf6 (see above)
```bash
node scripts/create2/deploy.js NexusViewer -a 0xcafeaaaA618809F4414fa9e2eB7Db9975419fa3F \
                                           -f 0xfac7011663910F75CbE1E25539ec2D7529f93C3F \
                                           -c '["0x01BFd82675DBCc7762C84019cA518e701C0cD07e","0xcafeaaf26278671fc393c60c09d936de2E3F6a1a","0xcafeaC01F6BA70567f5EA7dD4e114f8E5FDa5bf6"]' \
                                           -s 90662620 
                                           -b XX \
                                           -p 1
```
## Test plan

* add StakingViewer new methods fork tests
* add AssessmentViewer fork tests

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/assessment/AssessmentViewer.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  /// @notice Check if the stake of a member is locked
+  /// @param member The address of the member
+  /// @return stakeLocked Boolean indicating if the stake is locked
+  function isStakeLocked(address member) external view returns (bool stakeLocked) {
+
+    IAssessment _assessment = assessment();
+
+    if (nxmToken.isLockedForMV(member) > block.timestamp) {
```
Comment (roxdanila) [inline contracts/modules/assessment/AssessmentViewer.sol:51]: nit: this should be `>=`, for consistency

File: contracts/modules/assessment/AssessmentViewer.sol [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      return true; // NXM is locked for voting in governance
+    }
+
+    uint voteCount = _assessment.getVoteCountOfAssessor(member);
+    if (voteCount == 0) return false;
+
+    (,, uint32 timestamp,) = _assessment.votesOf(member, voteCount - 1);
```
Comment (roxdanila) [inline contracts/modules/assessment/AssessmentViewer.sol:58]: rename timestamp -> lastVoteTimestamp, for readability

File: contracts/modules/assessment/AssessmentViewer.sol [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    (, uint8 stakeLockupPeriodInDays,,) = _assessment.config();
+
+    uint stakeLockupExpiry = timestamp + stakeLockupPeriodInDays * 1 days;
+    stakeLocked = stakeLockupExpiry > block.timestamp;
```
Comment (roxdanila) [inline contracts/modules/assessment/AssessmentViewer.sol:62]: you used return false above, but assignment to the return value here. let's keep consistent one way or the other

File: test/fork/assessment-viewer.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    this.assessment = await ethers.getContractAt(abis.Assessment, addresses.Assessment);
+    this.nxm = await ethers.getContractAt('NXMToken', addresses.NXMToken);
+    this.individualClaims = await ethers.getContractAt(abis.IndividualClaims, addresses.IndividualClaims);
+  });
+
+  it('deploy AssessmentViewer', async function () {
+    this.assessmentViewer = await ethers.deployContract('AssessmentViewer', [this.master.address, this.nxm.address]);
+  });
+
+  it('getManagerPoolsAndRewards should return manager pools and rewards', async function () {
```
Comment (roxdanila) [inline test/fork/assessment-viewer.js:42]: wrong description

File: test/fork/assessment-viewer.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    const assessmentRewards = await this.assessmentViewer.getRewards(member);
+    expect(assessmentRewards.totalPendingAmountInNXM).to.be.instanceOf(BigNumber);
+    expect(assessmentRewards.withdrawableAmountInNXM).to.be.instanceOf(BigNumber);
+    expect(assessmentRewards.withdrawableUntilIndex).to.be.instanceOf(BigNumber);
+  });
```
Comment (roxdanila) [inline test/fork/assessment-viewer.js:47]: these checks don't add any value tbh - if we want to properly test these, we should write integration tests and check the actual values. same applies to all tests below

**Review by roxdanila — CHANGES_REQUESTED** [review]
Minor nits in the contracts. 
Tests need to be changed more significantly tho. Please avoid checks such as `expect(X).to.be.instanceOf(BigNumber);` - they don't add much value. Also I don't see fork tests as the best fit here. Ideally we stick to unit / integration tests, with proper setup and checking actual values.

File: contracts/modules/assessment/AssessmentViewer.sol (hunk 1 above)
Comment (rackstar) [reply contracts/modules/assessment/AssessmentViewer.sol]: dropped

File: test/fork/assessment-viewer.js (hunk 5 above)
Comment (rackstar) [reply test/fork/assessment-viewer.js]: converted fork tests to unit tests

File: test/integration/MCR/updateMCR.js [hunk 6]
```diff
@@ -41,6 +41,7 @@ async function updateMCRSetup() {
 
   const operator = await tk.operator();
   await setEtherBalance(operator, parseEther('10000000'));
+  await setEtherBalance(member1.address, parseEther('1000000'));
```
Comment (rackstar) [inline test/integration/MCR/updateMCR.js:44]: not sure why this integration is suddenly failing with "not enough funds" when ran along with the whole test suite.

```
  1) updateMCR
       increases desiredMCR:
     InvalidInputError: sender doesn't have enough funds to send tx. The max upfront cost is: 100000029026712406373968 and the sender's account only has: 9999999323294995263065

```

the test it does not seem to fail when ran on its own.

All the changes in this PR are confined within unit tests / viewer contracts so not sure how its affecting that particular integration test. but adding more funds to the address seems to fix it.

https://github.com/NexusMutual/smart-contracts/actions/runs/9870557531/job/27257740964

File: contracts/mocks/modules/Assessment/ASMockAssessment.sol [hunk 7]
```diff
@@ -0,0 +1,123 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity >=0.5.0;
+
+import {IAssessment} from "../../../interfaces/IAssessment.sol";
```
Comment (roxdanila) [inline contracts/mocks/modules/Assessment/ASMockAssessment.sol:5]: to follow the convention, create a new folder under mocks/modules called "AssessmentViewer" and move this file there and call it AVMockAssessment - see CoverViewer for an example

File: test/unit/StakingViewer/utils.js [hunk 8]
```diff
@@ -0,0 +1,10 @@
+const { ethers } = require('hardhat');
+
+const ONE_DAY_SECONDS = 24 * 60 * 60;
+
+async function calculateCurrentTrancheId() {
```
Comment (roxdanila) [inline test/unit/StakingViewer/utils.js:5]: this function already exists in fork/utils.js

utils across all tests deserve a revisit / cleanup for sure, but let's avoid duplicating this for now

File: contracts/mocks/modules/Assessment/ASMockAssessment.sol (hunk 7 above)
Comment (rackstar) [reply contracts/mocks/modules/Assessment/ASMockAssessment.sol:5]: oh, i see! i completely misunderstood how the naming convention works. thanks

File: test/unit/StakingViewer/utils.js (hunk 8 above)
Comment (rackstar) [reply test/unit/StakingViewer/utils.js:5]: I was hesitant to import something from fork dir to unit dir, hence I copied it over. but I see the point of avoiding dulpication

**Review by roxdanila — APPROVED** [review]

**Comment by rackstar** [comment]
marking as DRAFT for now until cover products changes are merged

**Pull request merged**